dicom_util/src/main/java/com/google/cloud/healthcare/DicomWebClientJetty.java

```diff
@@ -165,9 +160,125 @@ public void onData(Stream stream, DataFrame frame, Callback callback) {
+      if (credentials != null) {
+        credentials.getRequestMetadata();
+        requestFields.add(
+            HttpHeader.AUTHORIZATION, "Bearer " + credentials.getAccessToken().getTokenValue());
+      }
+
+      MetaData.Request request =
+          new MetaData.Request("DELETE", uri, HttpVersion.HTTP_2, requestFields);
+      HeadersFrame headersFrame = new HeadersFrame(request, null, true);
+
+      // Prepare the listener to receive the HTTP response frames.
+      final StringBuilder resultBuilder = new StringBuilder();
+      final CompletableFuture<Integer> responseCodeFuture = new CompletableFuture<>();
+      final CompletableFuture<Boolean> doneFuture = new CompletableFuture<>();
+
+      session.newStream(
+          headersFrame,
+          new FuturePromise<>(),
+          new Stream.Listener.Adapter() {
+            @Override
+            public void onReset(Stream stream, ResetFrame frame) {
+              doneFuture.complete(false);
+            }
+
+            @Override
+            public void onHeaders(Stream stream, HeadersFrame frame) {
+              if (frame.getMetaData() instanceof Response) {
+                responseCodeFuture.complete(((Response) frame.getMetaData()).getStatus());
+              }
+            }
+
+            @Override
+            public void onData(Stream stream, DataFrame frame, Callback callback) {
+              byte[] bytes = new byte[frame.getData().remaining()];
+              frame.getData().get(bytes);
+              resultBuilder.append(new String(bytes, StandardCharsets.UTF_8));
+
+              if (frame.isEndStream()) {
+                doneFuture.complete(true);
+              }
+
+              callback.succeeded();
+            }
+          });
+
+      doneFuture.get();
+      client.stop();
+
+      int httpStatus = responseCodeFuture.get();
+      if (httpStatus != HttpStatus.OK_200) {
+        String resp = resultBuilder.toString();
+        throw new DicomWebException(
+            "Http_" + httpStatus + ": " + resp, httpStatus, Status.ProcessingFailure);
+      }
+    } catch (Exception e) {
+      if (e instanceof DicomWebException) {
+        throw (DicomWebException) e;
+      }
+      throw new DicomWebException(e);
+    }
+  }
+
+  /**
+   * Reads the resource to figure out the study/series/instance UID path, deletes it and returns
+   * void.
+   *
+   * @param stream The unread stream containing the DICOM instance.
+   */
+  @Override
+  public void delete(InputStream stream) throws IDicomWebClient.DicomWebException {
+    try {
+      DicomInputStream dis = new DicomInputStream(stream);
+      Attributes attrs = dis.readDataset(-1, Tag.PixelData);
+      String instanceUrl =
+          String.format(
+              "%s/series/%s/instances/%s", // NOTE: /studies/ already in stowPath
+              attrs.getString(Tag.StudyInstanceUID, 0),
+              attrs.getString(Tag.SeriesInstanceUID, 0),
+              attrs.getString(Tag.SOPInstanceUID, 0));
```

Comment:
Nit: instead of defaulting to a 0 value and potentially deleting something we're not supposed to (instance with IDs of 0), can we throw an exception if these are not found?

Reply:
Resolved in latest commmit